### PR TITLE
[Cookbook Recipe of Using Rubric]

### PR DESCRIPTION
**Pull request by ruiqi-zhong**
# Rubric-based Grading for LLMs

- [`data.py`](./data.py) contains the definition for the datapoint class. Each datapoint consists of a conversation prefix and a list of rubric items.
- [`generate_data.py`](./generate_data.py) generates some example datapoint if you want to run our demo on addition.
- [`env.py`](./env.py) determines what each rollout will do. It will let the policy read the prefix, generate a response, ask a grader LLM to grade based on a list of rubric items, and finally provide a reward by summing the response of each grader.
- [`train.py`](./train.py) allows you train LLMs on any dataset saved in our format (specified in `data.py`). The default script will train on the addition task, whose data is generated by `generate_data.py`.
- [`prometheus_experimental.py`](./prometheus_experimental.py) contains a script to train the LLMs based on the rubrics from the [`prometheus-eval/Feedback-Collection`](https://huggingface.co/datasets/prometheus-eval/Feedback-Collection/viewer/default/train?row=0&views%5B%5D=train) dataset. It is experimental though -- even though the reward goes up, there is no guarantee that the model is actually better. We hope our script serves as a starting point, and more research is needed.

## A simple example of using a grader LLM with rubrics

We show how to use rubric-based LLM to provide reward on an addition task. E.g.

```
**User**: What's 233 + 100?
**Assistant**: 333
```

Usually, this could be graded by matching the number to the ground truth 333 without needing an LLLM. However, for pedagogical purposes we will grade the response using a language model with rubric. I.e. We will ask a language mode "Does the assistant answer 333?"

### Generate an example dataset

To run this, first generate a dataset:

```
python -m tinker_cookbook.recipes.rubric.generate_data
```

Then you will see two `jsonl` file generated, one for training, one for testing. For example, if you look into ` tinker_cookbook/example_data/example_rubric_train.jsonl`, each datapoint consists of
- a convo (the conversation prefix that the policy sees)
- rubric_items: a list of rubric items that specify what is a good item, how the grader should format the response, and how the grading result should be extracted.

```
{
  "convo": [
    {
      "role": "user",
      "content": "What is 4 + 5?"
    },
    {
      "role": "assistant",
      "content": "9"
    },
    {
      "role": "user",
      "content": "What is 122 + 12?"
    }
  ],
  "rubric_items": [
    {
      "rubric_str": "Does the chatbot correctly gets the answer 134?",
      "extraction_regex": "<score>(.*)</score>",
      "grader_output_format_instruction": "Please output your score between 0 and 1 wrapped in <score> ... </score>"
    }
  ]
}
```

### Debugging and Printing What Happens During Rollouts

Run
```
python -m tinker_cookbook.recipes.rubric.debug_env
```

You can see the message that the policy sees, its response, the grader input, and the grader output.

image

### An example training run

To train the LLM to add with a rubric-based LLM, run
```
python -m tinker_cookbook.recipes.rubric.train
```

You can see the reward quickly goes up.

image

### A more realistic dataset

We take the `prometheus-eval/Feedback-Collection` dataset from [hugingface](https://huggingface.co/datasets/prometheus-eval/Feedback-Collection/), which contains rubric to grade general chat responses. Run the following to kick off training:

```
python -m tinker_cookbook.recipes.rubric.prometheus_experimental
```

We can see that the reward climbs up steadily.

image

Note that this training recipe is experimental -- to make the performance better we may need to fine-tune the grader LLM as well. We hope our code serves as a starting for you to improve rubric-based grading for training LLMs!

### CODE REVIEW

File: tinker_cookbook/recipes/rubric/data.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    def _flatten_convo(self, convo: Conversation) -> str:
+        """
+        Convert the whole conversation (user's turns + assistant's turns) into a single string. E.g.
+        \n\nHuman: ....
+        \n\nChatbot: ...
+        \n\nHuman: ...
+        \n\nChatbot: ...
+        """
+        return "\n\n".join(
+            [f"{self.__convert_role(message['role'])}: {message['content']}" for message in convo]
```
Comment (joschu) [inline tinker_cookbook/recipes/rubric/data.py]: double underscore methods have a special meaning in python -- why not use single underscore?

File: tinker_cookbook/recipes/rubric/env.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import asyncio
+from tinker_cookbook import model_info
+from tinker_cookbook.recipes.rubric.data import (
+    RubricBasedDatapoint,
+    Rubric,
+    Conversation,
+    RubricDatapointListBuilder,
+)
+
+# ANSI color codes
```
Comment (joschu) [inline tinker_cookbook/recipes/rubric/env.py]: you can use the termcolor library

File: tinker_cookbook/recipes/rubric/data.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        convo: Conversation = [
+            {"role": "user", "content": item["orig_instruction"]},
+        ]
+
+        rubric_text = f"Your job is to evaluate the following: {item['orig_criteria']}. Your response should be a score between 1 to 5.\n"
+        rubric_text += "Here is the calibration for each score:\n"
+        for i in range(1, 6):
+            rubric_text += f"<score>{i}.0</score>: {item[f'orig_score{i}_description']}\n"
+
+        rubric_text += f"\nHere is a reference response that achieved a score of 5: {item['orig_reference_answer']}\n"
```
Comment (joschu) [inline tinker_cookbook/recipes/rubric/data.py]: some inconsistency in whether you put newline before or after line. how about just keeping a list of lines, and joining them at the end?

File: tinker_cookbook/recipes/rubric/debug_env.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    return datapoint
+
+
+async def main(datapoint: RubricBasedDatapoint):
+    policy_name = "meta-llama/Llama-3.1-8B-Instruct"
+    grader_name = "Qwen/Qwen3-30B-A3B-Instruct-2507"
+    service_client = tinker.ServiceClient()
+    policy = TinkerTokenCompleter(
+        sampling_client=service_client.create_sampling_client(base_model=policy_name),
+        max_tokens=64,
```
Comment (joschu) [inline tinker_cookbook/recipes/rubric/debug_env.py]: nit: pull this to the top of the function, so arbitrary parameter values are all together

File: tinker_cookbook/recipes/rubric/data.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        return "\n\n".join(
+            [f"{self.__convert_role(message['role'])}: {message['content']}" for message in convo]
+        )
+
+    def get_grader_prompt(self, convo: Conversation) -> Conversation:
+        """
+        Create a prompt for the grader to grade the conversation based on the rubric. The prompt should contain 1) the conversation to be graded, and 2) the rubric.
+        """
+
+        prompt = "I will show you 1) a conversation between a human and a chatbot, and 2) a rubric for grading the conversation. Please grade the conversation based on the rubric."
```
Comment (joschu) [inline tinker_cookbook/recipes/rubric/data.py]: this prompt doesn't say anything indicating that the grading should focus on the last assistant message, right? how about separating the prompt convo (which is fixed) from the completion (which the grader should look at)?

File: tinker_cookbook/recipes/rubric/README.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+### A more realistic dataset
+
+We take the `prometheus-eval/Feedback-Collection` dataset from [Hugging Face](https://huggingface.co/datasets/prometheus-eval/Feedback-Collection/), which contains rubrics to grade general chat responses. Run the following to kick off training:
+
+```
+python -m tinker_cookbook.recipes.rubric.prometheus_experimental
+```
+
+We can see that the reward climbs up steadily.
```
Comment (joschu) [inline tinker_cookbook/recipes/rubric/README.md:90]: it would be nice to show that the model is learning something meaningful

**Review by joschu — APPROVED** [review]
